pkg/reconciler/scheduling/namespace/namespace_reconcile_scheduling.go

```diff
@@ -0,0 +1,266 @@
+/*
+Copyright 2022 The KCP Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package namespace
+
+import (
+	"context"
+	"encoding/json"
+	"math/rand"
+	"strings"
+	"time"
+
+	"github.com/kcp-dev/logicalcluster"
+
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/types"
+	utilerrors "k8s.io/apimachinery/pkg/util/errors"
+	"k8s.io/klog/v2"
+	"k8s.io/kube-openapi/pkg/util/sets"
+
+	schedulingv1alpha1 "github.com/kcp-dev/kcp/pkg/apis/scheduling/v1alpha1"
+	workloadv1alpha1 "github.com/kcp-dev/kcp/pkg/apis/workload/v1alpha1"
+	locationreconciler "github.com/kcp-dev/kcp/pkg/reconciler/scheduling/location"
+)
+
+const removingGracePeriod = 10 * time.Second
+
+// placementSchedulingReconciler schedules a workload for this ns. It checks the current placement annotation on the ns,
+// and find all valid workloadclusters.
+type placementSchedulingReconciler struct {
```

Comment:
`// TODO: generalize this to consider the resource type specified in the Placement, i.e. not only workload clusters`

Reply:
I wonder, is this really a scheduling/namespace controller or should rather be workload/namespace? I think the later. Reason: this controller looks deep into WorkloadCluster object, e.g. will consider capacity and other details for its scheduling decision.

Reply:
yes, moved it to workload/namespace